framework/src/userobject/ElementSubdomainModifier.C

```diff
@@ -0,0 +1,353 @@
+#include "DisplacedProblem.h"
+
+#include "libmesh/parallel_algebra.h"
+#include "libmesh/parallel.h"
+#include "libmesh/dof_map.h"
+
+#include "libmesh/parallel_ghost_sync.h"
+#include "libmesh/mesh_communication.h"
+
+InputParameters
+ElementSubdomainModifier::validParams()
+{
+  InputParameters params = ElementUserObject::validParams();
+  params.addClassDescription(
+      "Modify element subdomain ID. This userobject only runs on the undisplaced mesh, and it will "
+      "modify both the undisplaced and the displaced mesh.");
+  params.addParam<bool>("apply_initial_conditions",
+                        true,
+                        "Whether to apply initial conditions on the moved nodes and elements");
+  params.addParam<BoundaryName>(
+      "moving_boundary_name",
+      "Boundary to modify when an element is moved. A boundary with the provided name will be "
+      "created if not already exists on the mesh.");
+  params.set<bool>("use_displaced_mesh") = false;
+  params.suppressParameter<bool>("use_displaced_mesh");
+  return params;
+}
+
+ElementSubdomainModifier::ElementSubdomainModifier(const InputParameters & parameters)
+  : ElementUserObject(parameters),
+    _displaced_problem(_fe_problem.getDisplacedProblem().get()),
+    _apply_ic(getParam<bool>("apply_initial_conditions")),
+    _moving_boundary_specified(isParamValid("moving_boundary_name"))
+{
+}
+
+void
+ElementSubdomainModifier::initialSetup()
+{
+  if (_moving_boundary_specified)
+  {
+    _moving_boundary_name = getParam<BoundaryName>("moving_boundary_name");
+    setMovingBoundaryName(_mesh);
+    if (_displaced_problem)
+      setMovingBoundaryName(_displaced_problem->mesh());
+  }
+}
+
+void
+ElementSubdomainModifier::setMovingBoundaryName(MooseMesh & mesh)
+{
+  // We only need one boundary to modify. Create a dummy vector just to use the API.
+  const std::vector<BoundaryID> boundary_ids = mesh.getBoundaryIDs({{_moving_boundary_name}}, true);
+  _moving_boundary_id = boundary_ids[0];
+  mesh.setBoundaryName(_moving_boundary_id, _moving_boundary_name);
+  mesh.getMesh().get_boundary_info().sideset_name(_moving_boundary_id) = _moving_boundary_name;
+  mesh.getMesh().get_boundary_info().nodeset_name(_moving_boundary_id) = _moving_boundary_name;
+}
+
+void
+ElementSubdomainModifier::initialize()
+{
+  _moved_elems.clear();
+  _moved_displaced_elems.clear();
+  _moved_nodes.clear();
+}
+
+void
+ElementSubdomainModifier::execute()
+{
+  // First, compute the desired subdomain ID for the current element.
+  SubdomainID subdomain_id = computeSubdomainID();
+
+  // If the current element's subdomain ID isn't what we want
+  if (subdomain_id != std::numeric_limits<SubdomainID>::max() &&
+      _current_elem->subdomain_id() != subdomain_id)
+  {
+    // Current element ID, used to index both the element on the displaced and undisplaced meshes.
+    dof_id_type elem_id = _current_elem->id();
+    // Change the element's subdomain
+    Elem * elem = _mesh.elemPtr(elem_id);
```

Comment:
`elem` will be different from `_current_elem`?  That being said, can we directly use `_current_elem` instead of `elem` in the following several places?

Reply:
But `_current_elem` is `const`, so I can't use it to directly change the subdomain ID.